web-api/src/storage/postgres.rs

```diff
@@ -414,6 +492,88 @@ impl storage::Interaction for Storage {
 
         Ok(())
     }
+
+    #[allow(clippy::too_many_lines)]
+    async fn update_interactions<F>(
+        &self,
+        user_id: &UserId,
+        updated_document_ids: &[&DocumentId],
+        mut update_logic: F,
+    ) -> Result<(), Error>
+    where
+        F: for<'a, 'b> FnMut(InteractionUpdateContext<'a, 'b>) -> &'b PositiveCoi + Send + Sync,
+    {
+        let mut tx = self.postgres.pool.begin().await?;
+
+        Database::acquire_user_coi_lock(&mut tx, user_id).await?;
+
+        let documents = self
+            .get_by_ids_with_transaction(updated_document_ids, Some(&mut tx))
+            .await?;
+
+        let now = Utc::now();
+
+        let mut document_map = documents
+            .iter()
+            .map(|d| (&d.id, d))
+            .collect::<HashMap<_, _>>();
+
+        let mut interests = Database::get_user_interests(&mut tx, user_id).await?;
+
+        let mut category_weight_diff_map = documents
+            .iter()
+            // yes this is intended to add an entry for `None` to
+            // if it appears on a document
```

Comment:
Please take a look after rebasing, we don't have `None` here anymore.

Reply:
thanks :+1: